### ISSUE TITLE
RuntimeException with Incorrect Advice

### ISSUE THREAD

**Issue by divil5000** [dotnet/runtime#100524]
### Description

On iOS, we receive unhandled exception telemetry on a daily basis reporting RuntimeExceptions with the following message:

Failed to marshal the Objective-C object 0x303ce1d80 (type: AppNamespace_OnscreenTileLayer). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance (because the type 'AppNamespace.OnscreenTileLayer' does not have a constructor that takes one NativeHandle argument).

But of course, we have exactly such a constructor:

```
public OnscreenTileLayer(ObjCRuntime.NativeHandle i)
	: base(i)
{
}
```

Obviously there's something wrong here either in the runtime or in the exception message being generated. I would be happy to be advised why this is likely to be occurring, so I can do something about it.

### Reproduction Steps

We have been unable to reproduce this exception in-house.

### Expected behavior

The expected behaviour is that this condition never arises, but if it does, the runtime should be able to create a managed instance from the constructor which clearly exists.

### Actual behavior

The exception is thrown despite the fact that a suitable constructor exists.

### Regression?

Not a regression; we've had these reports for years, spanning both the Xamarin framework (when it was looking for an IntPtr constructor) and now net7.0-iOS (where it's looking for a NativeHandle constructor).

### Configuration

net7.0-iOS

**Comment by divil5000** [comment]
This is a large, commercial and proprietary application in widespread use. I can't just provide all its source code here.

I was hoping you'd be able to look at the runtime code that produces the exception, and figure out how it could produce that exception when the condition reported is false.

**Comment by rolfbjarne** [comment]
> Without a repro sample it is really hard to tell what the problem could be. Could you provide the project which causes such exceptions?
> 
> @rolfbjarne could this be similar to what was reported here [xamarin/xamarin-macios#19579 (comment)](https://github.com/xamarin/xamarin-macios/issues/19579#issuecomment-1841228829) ?

No, that's unlikely, that's a rather peculiar implementation in the CoreBluetooth framework I haven't seen anywhere else.

@divil5000 do you call Dispose on any `OnscreenTileLayer` instances at any point? If so, try not disposing them to see if that makes a difference.

If not, it would really help to create a reproducible test case.

One trick that often helps in tracking down this particular type of problem is to run the GC in a loop in the background, by doing something like this in FinishedLaunching:

```cs
var thread = new Thread (() => 
{
    while (true) {
        Thread.Sleep (1000);
        GC.Collect ();
    }
});
thread.IsBackground = true;
thread.Start ();
```

and then automate the creation of `OnscreenTileLayer` somehow and try running the automation overnight.

**Comment by rolfbjarne** [comment]
> Failed to marshal the Objective-C object 0x303ce1d80 (type: AppNamespace_OnscreenTileLayer). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance (because the type 'AppNamespace.OnscreenTileLayer' does not have a constructor that takes one NativeHandle argument).
> 
> But of course, we have exactly such a constructor:

It could be that the linker is removing the constructor.

You can forcing the linker to keep it by adding the `[Preserve]` attribute:

```cs
[Preserve]
public OnscreenTileLayer(ObjCRuntime.NativeHandle i)
	: base(i)
{
}
```

Note that this may just trade one issue for another, if the managed `OnscreenTileLayer` instance that got collected by the GC had any managed state, that managed state will be gone ( event handlers are common, but any fields on the type will be cleared).

Generically the "Failed to marshal the Objective-C object" exception indicates a different problem, and that must be tracked down and fixed. 

See https://stackoverflow.com/a/13059140/183422 for a bit of related background info.

**Comment by rolfbjarne** [comment]
> Hi Rolf, yes we do call .Dispose on it. These are CALayer-derived objects and when we are done with them we effectively call:
> 
> `layer.RemoveFromSuperLayer(); layer.Dispose();`
> 
> To remove them from the screen then get rid of them. This would seem to be a pretty common pattern, especially as these layers are instantiated, displayed, and removed constantly (think map tiles as the user scrolls around).

My guess is that this happens:

1. You create the `OnscreenTileLayer` instance and add it to the view hierarchy.
2. You call RemoveFromSuperLayer + Dispose on the instance. This will disconnect the managed `OnscreenTileLayer` instance from the native (Objective-C) `OnscreenTileLayer` counterpart (which allows the managed GC to collect the managed instance if you don't reference it anywhere else, and the same thing for the Objective-C instance in native code).
3. For some reason iOS still has a reference to the native Objective-C instance, and decides to call a selector on it - and this is a selector you've overridden in managed code. At this point we need to find the managed instance corresponding with the native instance, but since they're disconnected (because you call Dispose), we can't find the managed instance. So we try to create a new one, and that fails with the exception you've seen.

> Is your recommendation still to simply not .Dispose them? What are the implications of this?

There are two implications:

* The `OnscreenTileLayer` will take longer to be collected by the garbage collector. Your app might end up using a bit more memory (until the GC runs).
* If there are any circular references (one example being the StackOverflow post I mentioned [earlier](https://stackoverflow.com/q/13058521/183422), your app will leak memory (the `OnscreenTileLayer` instances will never go away). This should be fairly easy to verify: run the app on device using Instruments, and monitor the memory usage. It should look like a bunch of peaks and valleys (the valleys being when the GC runs) - running the GC manually as I suggested below will likely yield more consistent results here.

If you're concerned about any of these two implications, I'd suggest trying to figure out why iOS calls methods on the `OnscreenTileLayer` instance after you've disposed it. One way to do this is to set flag instead of disposing, and then checking the flag in every overridden method  and check why iOS called the method in question.

> Is your recommendation still to simply not .Dispose them?

Our recommendation is to only call Dispose if you're certain iOS won't call any methods on the object anymore. The simplest way to ensure this is to not call Dispose (which typically works well unless the app uses a lot of memory very fast, such as photo/video/audio processing).

**Comment by divil5000** [comment]
Thank you for the detailed explanation. I think that because we override ActionForKey in the class, iOS must certainly be trying to call that selector after we have called .Dispose on the object, and that explains what we are seeing here. The exception message is still very misleading but from what you've written, it sounds like a plausible explanation.

The class in question doesn't have any events and it's very lightweight with no references at all, so I think circular references are not a problem here.

It sounds like my only option is to simply no longer call .Dispose, and just allow the GC to do its work in due course, would you agree? I will monitor this to see if we appear to be leaking as a result.

**Comment by rolfbjarne** [comment]
> It sounds like my only option is to simply no longer call .Dispose, and just allow the GC to do its work in due course, would you agree?

Yes, that sounds like a good plan.

**Comment by ivanpovazan** [comment]
Thanks @rolfbjarne!

@divil5000 I will close the issue now but please fill free to reopen it if you encounter any further problems.
